libfabric.map.in

```diff
@@ -7,6 +7,11 @@ FABRIC_1.0 {
 		fi_version;
 		fi_strerror;
 		fi_tostr;
+		fi_log_set_stream_warn;
+		fi_log_set_stream_trace;
+		fi_log_set_stream_info;
+		fi_log_set_stream_debug;
+		fi_log_set_func;
```

Comment:
These cannot go here.  This breaks the ABI.  New functions must go into a new map area, DEFAULT_SYMVER needs to be updated, existing API calls modified, etc.  See include/ofi_abi.h for a discussion on this.

The existing API has 2 functions for logging, fi_log and fi_log_enabled.  Those make use of an enum for the log level.  We do not want to add new APIs that replace that enum by breaking it into single functions per log level.  That is, if we add a new enum value, the new APIs are not updated and require another API update.

If I understand correctly, the intent is to replace stderr for any specific log message.  That could be done using some sort of more generic call such as fi_log_config().  If we add such a call, we need to design into it what part of the log message can be configured.  The commit message talks about adding time stamps and other items to the log message, but that is not what the changes are doing.  A more generic configure call would allow this.

The fi_log APIs are provider facing, not application facing.  Trying to allow an application to configure logging will be difficult.  Logging is likely to occur prior to an application trying to re-configure it.  As a result, current log configurations are controlled through environment variables.

There needs to be a clear description of *why* the changes are being proposed.  That's essential in order to analyze the best option for any possible changes.

Reply:
> These cannot go here. This breaks the ABI. New functions must go into a new map area, DEFAULT_SYMVER needs to be updated, existing API calls modified, etc. See include/ofi_abi.h for a discussion on this.
> 

Ok, understood for the ABI rules, thanks.

> The existing API has 2 functions for logging, fi_log and fi_log_enabled. Those make use of an enum for the log level. We do not want to add new APIs that replace that enum by breaking it into single functions per log level. That is, if we add a new enum value, the new APIs are not updated and require another API update.
> 

May be a single function to set stream with the level as a parameter ?

> If I understand correctly, the intent is to replace stderr for any specific log message. That could be done using some sort of more generic call such as fi_log_config(). If we add such a call, we need to design into it what part of the log message can be configured. The commit message talks about adding time stamps and other items to the log message, but that is not what the changes are doing. A more generic configure call would allow this.
> 

With the ABI to change the function log to an external one, adding time-stamp/thread-id/.... can be done in there.

> The fi_log APIs are provider facing, not application facing. Trying to allow an application to configure logging will be difficult. Logging is likely to occur prior to an application trying to re-configure it. As a result, current log configurations are controlled through environment variables.
> 

This change do not reconfigure current logging, it just allows to enhance and redirect it.

> There needs to be a clear description of _why_ the changes are being proposed. That's essential in order to analyze the best option for any possible changes.

It is essential because presently, stderr logs from OFI/libfabric can garble the output of an application using it. When and intermediate layer uses specific log file, causing OFI logs to be out of scope (with no timestamp nor thread-id). This is particularly useless and annoying with debug mode.

Reply:
The commit message needs to capture the detailed reasons for the change.  The PR should also only include the final patches to merge.  You can force push changes as the patches are updated.

Adding time stamps and thread id values to the libfabric log messages should be doable.  Those can be added as separate patches.

Reply:
> The commit message needs to capture the detailed reasons for the change.

Ok.

> Adding time stamps and thread id values to the libfabric log messages should be doable.

It is much efficient to provide a way for applications/layers using libfabric to add whatever they need to current msg format, like being done in this patch.

Also, concerning libfabric's ABI specials you have derailed before, I have a question, how can an external application use both fi_log()/fi_log_enabled() ABIs since associated include/rdma/providers/fi_log.h defining their prototype is not part of the provided header files within distro ?

Reply:
This patch *requires* applications to change to be of use.  That's definitely not more efficient.

The log APIs are targeting externally linked providers, not applications.  They are specifically not installed for that reason.  It's usual for a library to expose APIs for an application to change its log messages.  I've never seen another library expose something like that.